[Sema] Fix a couple of pattern type-checking bugs

hamishknight opened the pull request:
Avoid checking a switch for exhaustiveness in a multi-statement closure if the solution failed to apply, and ensure we always run redeclaration checking for bound pattern vars.

rdar://105781521
Resolves #63750

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

ahoppen reviewed (approved):
LGTM. Thanks

CodaFi commented on lib/Sema/CSSyntacticElement.cpp:
```diff
@@ -1738,9 +1738,10 @@ class SyntacticElementSolutionApplication
-
+    if (!hadError) {
+      TypeChecker::checkSwitchExhaustiveness(
+          switchStmt, context.getAsDeclContext(), limitExhaustivityChecks);
```
I would just check exhaustiveness anyways but throw the limited mode switch so folks don’t have to build twice to see _something_.

hamishknight replied:
Makes sense, thanks!

hamishknight committed:
[CS] Perform limited exhaustiveness checking if we couldn't apply the solution

Currently ExprPatterns are type-checked during
CSApply. As such, they can cause solution
application to fail with a pattern that isn't
well-formed. I'm planning on moving their
type-checking into the solver, but until then,
lets only do limited exhaustiveness checking for
the switch if there was an error.

rdar://105781521

hamishknight committed:
[Sema] Diagnose duplicate pattern vars in pattern resolution

Unify the duplicate pattern var checking in
pattern resolution such that it is always called
no matter how we end up type-checking the pattern.
This avoids incorrectly allowing duplicate pattern
vars to compile for patterns as a part of
multi-statement closures.

hamishknight commented:
@swift-ci please test
